### ISSUE TITLE
`redundant_closure` makes incorrect suggestion

### ISSUE THREAD

**Issue by lapla-cogito** [rust-lang/rust-clippy#13895]
### Summary

`redundant_closure` lint sometimes makes incorrect suggestion in a `no_std` environment.

### Reproducer

I tried this code:

```rust
#![no_std]
extern crate alloc;
use alloc::vec;
use alloc::vec::Vec;

fn main() {
    let _: Option<Vec<u8>>= true.then(|| vec![]);
}
```

I expected to see this happen:

Suggest code that compiles successfully in a `no_std` environment, e.g. using `alloc::vec::Vec`, or not linting.

Instead, this happened:

The following was suggested, which fails to compile:
```rust
#![no_std]
extern crate alloc;
use alloc::vec;
use alloc::vec::Vec;

fn main() {
    let _: Option<Vec<u8>>= true.then(std::vec::Vec::new);
}
```

### Version

```text
rustc 1.85.0-nightly (dd84b7d5e 2024-12-27)
binary: rustc
commit-hash: dd84b7d5eec3c20d7fcd13e6450af029d3cece9d
commit-date: 2024-12-27
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by lapla-cogito** [comment]
This part is the cause:
https://github.com/rust-lang/rust-clippy/blob/c8ba3e1d7a92a591e01c13df79a868fff8876534/clippy_lints/src/eta_reduction.rs#L107-L115
